<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE html
     PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN"
     "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en">
<head>
<title>LingMed: LingPipe for Entrez-Pubmed</title>
<meta http-equiv="Content-type"
      content="application/xhtml+xml; charset=utf-8"/>
<meta http-equiv="Content-Language"
      content="en"/>
<link href="lp-site.css"
      type="text/css"
      rel="stylesheet"
      title="lp-site"
      media="screen,projection,tv"/>
<link href="lp-site-print.css"
      title="lp-site"
      type="text/css"
      rel="stylesheet"
      media="print,handheld,tty,aural,braille,embossed"/>
</head>

<body>
<div id="header">
<h1 id="product">Downloading LingMed</h1><h1 id="pagetitle">Getting Started</h1>
<a id="logo" href="http://alias-i.com/">
<img src="http://alias-i.com/lingpipe/web/img/logo-small.gif" alt="alias-i logo"/>
</a>
</div><!-- head -->

<div id="navig">
<!-- set class="current" for current link -->
<ul>
<li><a href="../read-me.html">lingmed</a>
<ul>
<li><a class="current" href="getting_started.html">getting started</a>
<ul>
<li><a href="#ant">using ant</a></li>
<li><a href="#ncbi">ncbi downloads</a></li>
<li><a href="#indexes">building indexes</a></li>
</ul>
<li><a href="dao.html">search and indexing</a>
<li><a href="medline.html">medline processing</a></li>
<li><a href="lingblast.html">lingblast</a></li>
<li><a href="genelinkage.html">putting it together: genelinkage</a></li>
</ul>
</li>
<li><a href="http://alias-i.com/lingpipe/index.html">lingpipe</a></li>
</ul>
</div><!-- navig -->

<div id="content" class="content">

<h2>LingMed download</h2>

<p>
If you are here you're obviously managed to download LingMed from the
Alias-i <a href="http://alias-i.com/lingpipe/web/sandbox.html">Development Sandbox</a>.
Remember to check back periodically for updates.
</p>

<a name="ant"><h2>Using Ant</h2></a>

<p>
<a href="http://ant.apache.org/">Apache Ant</a> is used to build and run LingMed applications.
The LingMed programs use command line arguments to get configuration information.  
Using Ant targets allows us to fully specify these arguments in the build files,
and then invoke them with the command: "ant <i>targetname</i>".
This saves a lot of typing.
If you're familiar with Ant, you can skip the rest of this section,
which explains how to modify these files for your installation.
</p>

<p>
The top-level LingMed directory contains a set of ant configuration files:
</p>
<ul>
<li>build.lingmed.properties - local configuration variables</li>
<li>build.xml - the default build file, runs core tasks</li>
<li>build_common.xml - common tasks and macros for common tasks</li>
<li>build_dev.xml - tasks used for testing and development</li>
<li>build_prod.xml - core tasks for a production environment</li>
<li>build_web.xml - targets that launch the genelinkage webapp</li>
</ul>


<p>
The default build file, and the dev, prod, and web files 
all start by accessing build.lingmed.properties and build_common.xml:
</p>

<pre class="code">
 &lt;property file="build.lingmed.properties"/&gt;
 &lt;import file="build_common.xml"/&gt;
</pre>

<p>
The build.lingmed.properties file contains properties in
the Java key=value format.  These properties are referenced
in the build xml files as ${key}.
For example, the location of lingpipe jar file is specified
in build.lingmed.properties as
"lingpipe-jar=lib/lingpipe-3.7.0.jar".
There are a few properties that you will want to change
according to your installation, including paths to the
directories for NLM distribution files and Lucene indexes.
</p>

<p>
The ant -f flag is used to specify a particular build file,
else ant will use the file <code>build.xml</code>.
This file contains the basic LingMed infrastructure targets.
This is all you need to build the indexes for each Entrez Gene
database and run them as search servers.
The build_dev.xml is used during applications development.
The build_prod.xml is for use in a production server environment,
which lacks a JDK, and it only contains targets to build and
run the LingMed infrastructure components.
</p>

<p>
Because of this proliferation of build files, we have used
Ant macros, which are defined in <code>build_common.xml</code>.
Macros are also useful for invoking LingMed programs which
take a number of installation-dependent arguments and one
task-specific argument, e.g. starting a remote search service,
or handling MEDLINE distribution files, which may be either
baseline or updates files.
</p>

<p>
Here is the macro for the LingMed DownloadMedline program:
</p>

<pre class="code">
&lt;macrodef name="downloadMedline"&gt;
  &lt;attribute name="repository"/&gt;
  &lt;attribute name="target"/&gt;
  &lt;attribute name="sleep"/&gt;
  &lt;sequential&gt;
    &lt;java classname="com.aliasi.lingmed.medline.DownloadMedline"
        fork="true"&gt;
      &lt;classpath refid="classpath.standard"/&gt;
      &lt;arg value="-domain=ftp.nlm.nih.gov"/&gt;
      &lt;arg value="-user=anonymous"/&gt;
      &lt;arg value="-password=${medline.pwd}"/&gt;
      &lt;arg value="-maxTries=${medline.download.retries}"/&gt;
      &lt;arg value="-repositoryPath=@{repository}"/&gt;
      &lt;arg value="-targetDir=@{target}"/&gt;
      &lt;arg value="-sleep=@{sleep}"/&gt;
    &lt;/java&gt;
  &lt;/sequential&gt;
&lt;/macrodef&gt;
</pre>

<p>
The MEDLINE distribution files come in two flavors: baseline and updates.
The NLM ftp repository keeps these in separate directories, and LingMed
mirrors this directory structure.  The LingMed program DownloadMedline
is used to download both kinds of files.  The macro definition allows us
to specify the names of the NLM repository and the local directory
in the Ant target as attributes, which are referenced by "@{attribute name}".  
Other configuration properties are the same,
and so these are referenced directly in the macro.
Here are the targets which use it in default build.xml file:
</p>

<pre class="code">
&lt;target name="downloadMedlineBaseline"
        description="download MEDLINE baseline distribution from NLM"
        depends="jar"&gt;
  &lt;downloadMedline repository="${medline.baseline.repository}"
                   target="${medline.baseline.dir}"
                   sleep="-1"/&gt;
&lt;/target&gt;

&lt;target name="downloadMedlineUpdates"
        description="download MEDLINE update files from NLM"
        depends="jar"&gt;
  &lt;downloadMedline repository="${medline.updates.repository}"
                   target="${medline.updates.dir}"
                   sleep="60"/&gt;
&lt;/target&gt;
</pre>

<a name="ncbi"><h2>Getting the NCBI Distribution Files</h2></a>

<p>
We have written a program to download the MEDLINE distribution files
from NLM.  All other databases are distributed as a single file, and
can be downloaded using your favorite ftp client or web browser.
</p>

<h3>MEDLINE Distribution Files</h3></a>

<p>
MEDLINE is distributed via FTP download from the
United States National Library of Medicine (NLM).
MEDLINE data is free for most purposes, but requires a 
<a href="http://www.nlm.nih.gov/databases/leased.html">
license agreement</a>.
You must register your IP address for downloading with the NLM,
so you'll need a fixed IP address to keep up with MEDLINE.
</p>

<p>
MEDLINE is distributed in 2 parts:  
the baseline distribution,
and a set of updates files.
The differences between the baseline and update files are:
</p>

<p>
<ul>
<li> 
All files in the baseline distribution become available at once.
Baseline files contain unique citation entries.
These files can be processed in any order, or in parallel.
</li>
<li> 
A new updates file is released once a day (on most weekdays).
Updates files contain new citation entries as well as
revisions of existing entries.  Updates files may also
contain instructions to delete existing entries.
In order to maintain a single, coherent Lucene index
over MEDLINE, each update must be processed in the order
in which it is released.
</li>
</ul>
</p>
<p>
The size of the downloaded 2009 baseline distribution is 8.7G.
The total size of all updates files distributed in 2008 was just under 2G,
so allocate disk space accordingly.
</p>

<p>
The NLM repository contains seperate directories for 
the baseline and updates files. 
The file names encode the year and 
the order in which the files are released.
For 2009, the baseline files are named 
<code>medline2009n0001.xml.gz</code>
through 
<code>medline2009n0593.xml.gz</code> 
and 
the first update file is named
<code>medline2009n0594.xml.gz</code>.
The 4 digit number preceeding the suffix <code>.xml</code>
is the ordering.
and this number is incremented with each subsequent update file.
</p>

<p>
We use the LingMed program DownloadMedline to fetch these files from NLM.
The <code>build.xml</code> file defines a target for 
downloading each kind of file: 
<code>downloadMedlineBaseline</code>
and <code>downloadMedlineUpdates</code>.
The properties that may need to be customized are:
</p>

<ul>
<li>medline.pwd - password for anonymous ftp, typically your e-mail</li>
<li>nlm.data - base directory for all nlm distribution files</li>
<li>medline.dir - if different than ${nlm.data}/medline/dist/2009</li>
<li>medline.baseline.dir - if different than "${medline.dir}/baseline"</li>
<li>medline.updates.dir - if different than "${medline.dir}/updates"</li>
</ul>

<p>
First download the baseline distribution files:
</p>
<pre class="code">
&gt; ant downloadMedlineBaseline
</pre>
<p>
This may take some time, depending the network.
Then download the updates files:
</p>
<pre class="code">
&gt; ant downloadMedlineUpdates
</pre>
<p>
The DownloadMedline program takes a command line argument <code>sleep</code>.
If this argument is a positive interger value N, the program will run perpertually,
sleeping N minutes between download attempts.  The default sleep value for
the Ant target <code>downloadMedlineUpdates</code> is 60,
which means that the program will check for new updates files once an hour.
To change this, edit the target definition in your build file.
</p>

<h3>Entrez Gene Distribution Files</h3></a>
<p>
The EntrezGene database is available for FTP download from the NCBI website.
The FTP repository is:
<a href="ftp://ftp.ncbi.nih.gov/gene/DATA/ASN_BINARY/">ftp.ncbi.nih.gov/gene/DATA/ASN_BINARY</a>.
Gene data is organized into subdirectories for different classes of species (e.g. mammalia),
one data file per species. 
Entrez Gene data are stored as compressed binary Entrezgene-Set ASN.1 filees 
and have the suffix &quot;.ags.gz&quot;.
Once downloaded, the files must be converted to .xml
using the 
<a href="ftp://ftp.ncbi.nih.gov/asn1-converters/documentation/gene2xml.txt">gene2xml</a> program, 
also distributed by NCBI via FTP download from the URL:
<a href="ftp://ftp.ncbi.nih.gov/asn1-converters">ftp.ncbi.nih.gov/asn1-converters</a>.
</p>

<p>
The distribution for each species is distributed as a fixed dataset,
and each new release replaces the previous release.
There is no versioning information on the release, 
and we have not figured out how to automatically determine
when a new version should be downloaded and processed.
Instead, we periodically download a release, process it,
and replace the existing EntrezGene index.
</p>

<p>
The URL for downloading the human genome is:
<a href="ftp://ftp.ncbi.nih.gov/gene/DATA/ASN_BINARY/Mammalia/Homo_sapiens.ags.gz">
ftp://ftp.ncbi.nih.gov/gene/DATA/ASN_BINARY/Mammalia/Homo_sapiens.ags.gz</a>.
We convert the file to XML and then strip out the DTD declaration
because the EntrezGene DTDs don't validate properly:
</p>

<pre class="code">
&gt; linux.gene2xml -b &lt;Homo_sapiens.ags | grep -v 'NCBI_Entrezgene.dtd' &gt; Homo_sapiens.xml
</pre>

<p>
Finally we move or copy this file to the location specified by
the buildfile property <code>entrezgene.dist.file</code>.
</p>

<h3>Homologene Distribution Files</h3></a>

<p>
Homologene is distributed as a compressed XML file.
Like EntrezGene, Homologene is distributed as one dataset,
and each new release replaces the previous release.
It can be downloaded from NCBI vis the URL:
<a href="ftp://ftp.ncbi.nih.gov/pub/HomoloGene/current/homologene.xml.gz">
ftp://ftp.ncbi.nih.gov/pub/HomoloGene/current/homologene.xml.gz</a>.
</p>

<h3>OMIM Distribution Files</h3></a>

<p>
The OMIM database is available for FTP download from
<A href="ftp://ftp.ncbi.nih.gov/repository/OMIM/omim.txt.Z">
ftp.ncbi.nih.gov/repository/OMIM/omim.txt.Z</A>
It is a compressed ASCII text file
containing a set of records.
Each record contains a number of fields.
Records begin with the header: 
<code>*RECORD*</code> on a separate line.
Fields begin with the header: 
<code>*FIELD* &lt;FIELD KEY&gt;</code> also on a separate line.
</p>

<a name="indexes"><h2>Building the indexes</h2></a>

<h3>MEDLINE</h3>

<p>
The LingMed program IndexMedline is used to build 
a Lucene index from the MEDLINE distribution files.
We build the index in two steps:  
first we process all the baseline files into a new index;
then we add the updates files to it.
The Ant task <code>indexMedlineBaseline</code> does the step one:
<pre class="code">
&gt; ant indexMedlineBaseline
</pre>
<p>
Building the index may take some time, depending on your machine.
You will need at least 800Mb memory, and at least 170GB of disk space.
The searchable Lucene index for the MEDLINE 2009 baseline distribution
is 84GB.
After the <code>indexMedlineBaseline</code> task has completed successfully,
index the updates files:
</p>
<pre class="code">
&gt; ant indexMedlineUpdates
</pre>

<p>
The IndexMedline program takes a command line argument <code>sleep</code>,
(like the DownloadMedline program does).
If this argument is a positive interger value N, the program will run perpertually.
The default sleep value for the Ant target <code>indexMedlineUpdates</code> is 60.
By running both the Ant tasks <code>downloadMedlineUpdates</code> and
<code>indexMedlineUpdates</code> this way,
updates files are downloaded and indexed in a timely fashion,
so that the local Lucene repository is in sync
with the on-line PubMed citation index.
</p>

<h3>EntrezGene, Homologene, and OMIM</h3>

<p>
The <code>build.xml</code> file contains tasks to create Lucene indexes
for each of these databases:  <code>indexEntrezGene</code>;
<code>indexHomologene</code>; <code>indexOMIM</code>.
</p>

<p>
To include some but not all species in the EntrezGene index, you should
download the individual species files 
(e.g.: 
<a href="ftp://ftp.ncbi.nih.gov/gene/DATA/ASN_BINARY/Mammalia/Mus_musculus.ags.gz">
ftp://ftp.ncbi.nih.gov/gene/DATA/ASN_BINARY/Mammalia/Mus_musculus.ags.gz</a>
for the mouse genome), 
and then run the <code>indexEntrezGene</code> task,
setting the "${entrezgene.dist.file}" property accordingly.
</p>

<p><em>Note:</em> take care not to run these indexing tasks
on the same distribution file more than once - this will result
in duplicate entries in the Lucene index.
</p>

</div><!-- content -->

<div id="foot">
<p>
&#169; 2003&ndash;2006 &nbsp;
<a href="mailto:lingpipe@alias-i.com">alias-i</a>
</p>
</div>

</body>

</html>


