<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE html
     PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN"
     "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en">
<head>
<title>LingMed: LingPipe for Entrez-Pubmed</title>
<meta http-equiv="Content-type"
      content="application/xhtml+xml; charset=utf-8"/>
<meta http-equiv="Content-Language"
      content="en"/>
<link href="lp-site.css"
      type="text/css"
      rel="stylesheet"
      title="lp-site"
      media="screen,projection,tv"/>
<link href="lp-site-print.css"
      title="lp-site"
      type="text/css"
      rel="stylesheet"
      media="print,handheld,tty,aural,braille,embossed"/>
</head>

<body>
<div id="header">
<h1 id="product">LingMed</h1><h1 id="pagetitle">Entrez Databases</h1>
<a id="logo" href="http://alias-i.com/">
<img src="http://alias-i.com/lingpipe/web/img/logo-small.gif" alt="alias-i logo"/>
</a>
</div><!-- head -->

<div id="navig">
<!-- set class="current" for current link -->
<ul>
<li><a href="../read-me.html">lingmed</a>
<ul>
<li><a href="dao.html">search and indexing</a></li>
<li><a href="medline.html">medline processing</a></li>
<li><a class="current" href="more_entrez.html">entrez databases</a>
<ul>
<li><a href="#entrezgene">entrez gene</a></li>
<li><a href="#homologene">homologene</a></li>
<li><a href="#omim">OMIM</a></li>
</ul>
</li>
<li><a href="lingblast.html">lingblast</a></li>
<li><a href="genelinkage.html">putting it together: genelinkage</a></li>
</ul>
</li>
<li><a href="http://alias-i.com/lingpipe/index.html">lingpipe</a></li>
</ul>
</div><!-- navig -->

<div id="content" class="content">

<h2>Overview</h2>

<p>
This section covers the LingMed tools and packages used for
search and indexing of the following 
<a href="http://www.ncbi.nlm.nih.gov/Database/">Entrez databases</a>:
<ul>
<li><a href="http://www.ncbi.nlm.nih.gov/sites/entrez?db=gene">Entrez Gene</a>
- gene-specific information</li>
<li><a href="http://www.ncbi.nlm.nih.gov/sites/entrez?db=homologene">HomoloGene</a>
- eukaryotic homology groups.</li>
<li><a href="http://www.ncbi.nlm.nih.gov/sites/entrez?db=omim">
OMIM: Online Mendelian Inheritance in Man</a>
- a compendium of human genes and genetic phenotypes.</li>
</ul>
</p>


<a name="entrezgene"><h2>Entrez Gene</h2></a>

<p>
The package 
<a href="api/com/aliasi/lingmed/entrezgene/package-summary.html"><code>com.aliasi.lingmed.entrezgene</code></a>
contains tools for creating and searching a Lucene index over Entrez Gene datasets.
It handles the mapping between an Entrez Gene database entry and structured Java objects of type
<a href="api/com/aliasi/lingmed/entrezgene/EntrezGene.html"><code>EntrezGene</code></a>.
</p>

<h3>Downloading Entrez Gene</h3>

<p>
The EntrezGene database is available for FTP download from the
<a href="http://www.ncbi.nlm.nih.gov/">National Center for Biotechnology Information</a>
(NCBI) website.
The FTP repository is:
<a href="ftp://ftp.ncbi.nih.gov/gene/DATA/ASN_BINARY/">ftp.ncbi.nih.gov/gene/DATA/ASN_BINARY</a>.
Gene data is organized into subdirectories for different classes of species (e.g. mammalia),
one data file per species. 
Entrez Gene data are stored as compressed binary Entrezgene-Set ASN.1 files 
and have the suffix &quot;.ags.gz&quot;.
Once downloaded, the files must be converted to .xml
using the 
<a href="ftp://ftp.ncbi.nih.gov/asn1-converters/documentation/gene2xml.txt">gene2xml</a> program, 
also distributed by NCBI via FTP download from the URL:
<a href="ftp://ftp.ncbi.nih.gov/asn1-converters">ftp.ncbi.nih.gov/asn1-converters</a>.
</p>

<p>
The distribution for each species is distributed as a fixed dataset,
and each new release replaces the previous release.
There is no versioning information on the release, 
and we have not figured out how to automatically determine
when a new version should be downloaded and processed.
Instead, we periodically download a release, process it,
and replace the existing EntrezGene index.
</p>

<p>
The URL for downloading the human genome is:
<a href="ftp://ftp.ncbi.nih.gov/gene/DATA/ASN_BINARY/Mammalia/Homo_sapiens.ags.gz">
ftp://ftp.ncbi.nih.gov/gene/DATA/ASN_BINARY/Mammalia/Homo_sapiens.ags.gz</a>.
We convert the file to XML and then strip out the DTD declaration
because the EntrezGene DTDs don't validate properly:
</p>

<pre class="code">
&gt; linux.gene2xml -b &lt;Homo_sapiens.ags | grep -v 'NCBI_Entrezgene.dtd' &gt; Homo_sapiens.xml
</pre>

<h3>Indexing Entrez Gene</h3>

<p>
The program
<a href="api/com/aliasi/lingmed/entrezgene/IndexEntrezGene.html"><code>IndexEntrezGene</code></a>
processes a file of XML-formatted EntrezGene entries 
and adds them to a Lucene index on the local filesystem.
It takes the following arguments:
</p>

<p>
<ul>
<li>
<b>index</b>
&mdash;  Path to the Lucene index file. 
</li>
<li>
<b>distFile</b>
&mdash; Path to the XML file.
</li>
</ul>
</p>

<p>
The LingMed 
<a href="../build.xml">build.xml</a> file
defines properties and tasks to run this program:
</p>

<pre class="code">
&lt;property name="entrezgene.lucene.index.dir"
          value="testluceneEntrezGene"/&gt;
&lt;property name="entrezgene.dist.file"
          value="C:\\aliasi\data\entrezgene\Homo_sapiens.xml"/&gt;
...
&lt;target name="indexEntrezGene"
        description="create Lucene EntrezGene index"
        depends="jar"&gt;
  &lt;java classname="com.aliasi.lingmed.entrezgene.IndexEntrezGene"
        maxMemory="800M"
        fork="true"&gt;
    &lt;classpath refid="classpath.standard"/&gt;
    &lt;arg value="-index=${entrezgene.lucene.index.dir}"/&gt;
    &lt;arg value="-distFile=${entrezgene.dist.file}"/&gt;
  &lt;/java&gt;
&lt;/target&gt;
</pre>

<p>
<code>IndexEntrezGene</code> builds a new Lucene index from the Entrez Gene XML file.
The XML is parsed into its constituent entries by a
<a href="http://alias-i.com/lingpipe/docs/api/com/aliasi/corpus/Parser.html">
<code>com.aliasi.corpus.Parser&lt;H extends Handler&gt;</code></a>, which provides SAX-style parsing
of content from an input source or character sequence 
and passing extracted events to a content handler of type
<a href="http://alias-i.com/lingpipe/docs/api/com/aliasi/corpus/ObjectHandler.html">
<code>com.aliasi.corpus.ObjectHandler&lt;E&gt;</code></a>.
The class
<a href="api/com/aliasi/lingmed/entrezgene/EntrezParser.html">
<code>EntrezParser</code></a> provides the Entrez Gene specific parser.
<code>EntrezParser</code> extends
<a href="http://alias-i.com/lingpipe/docs/api/com/aliasi/corpus/XMLParser.html">
<code>com.aliasi.corpus.XMLParser&lt;H extends Handler&gt;</code></a>.



</p>

<p>
<code>IndexEntrezGene</code> contains a static class called 
<code>EntrezGeneIndexer</code> that implements <code>ObjectHandler&lt;EntrezGene&gt;</code>
and which does the work of building the index.
The handler converts
<a href="api/com/aliasi/lingmed/entrezgene/EntrezGene.html"><code>EntrezGene</code></a> objects
to a Lucene <code>Document</code> using an
<a href="api/com/aliasi/lingmed/entrezgene/EntrezGeneCodec.html"><code>EntrezGeneCodec</code></a>,
and adds it to the Lucene index.
The necessary references to the <code>EntrezGeneCodec</code> and the 
<code>org.apache.lucene.index.IndexWriter</code> used by the <code>handle</code> method
are passed into the <code>EntrezGeneIndexer</code>'s constructor.
</p>

<pre class="code">
static class EntrezGeneIndexer implements ObjectHandler&lt;EntrezGene&gt; {
    final IndexWriter mIndexWriter;
    final EntrezGeneCodec mGeneCodec;

    public EntrezGeneIndexer(IndexWriter indexWriter, EntrezGeneCodec codec) {
        mIndexWriter = indexWriter;
        mGeneCodec = codec;
    }
</pre>

<p>
Because the <code>IndexEntrezGene</code> program is used to create a new verison
of the index, and because the XML file
consists of a set of unique <code>EntrezGene</code> entries,
there is no need to handle deletes from the index.
Therefore the <code>handle</code> method is straightforward:
</p>

<pre class="code">
public void handle(EntrezGene eg) {
    ...
    if (eg.isStatusLive() && eg.isTypeGene()) {
       ...
       Document doc = mGeneCodec.toDocument(eg);
       try { 
           mIndexWriter.addDocument(doc);  
       } catch (IOException ioe) {
           Logger.getLogger(IndexEntrezGene.class).warn("Exception indexing EntrezGene: " + ioe);
       }
    ...
</pre>

<p>
<i>Note:</i>If this command is run more than once over an XML file and the
Lucene index already exists, then the resulting index will contain duplicate
entries.
</p>


<h3>Searching Entrez Gene</h3>

<p>
The interface
<a href="api/com/aliasi/lingmed/entrezgene/EntrezGeneSearcher.html"><code>EntrezGeneSearcher</code></a>
defines the search functionality for Lucene index over Entrez Gene.
</p>

<a name="homologene"><h2>HomoloGene</h2></a>


<p>
The package 
<a href="api/com/aliasi/lingmed/homologene/package-summary.html"><code>com.aliasi.lingmed.homologene</code></a>
contains tools for creating and searching a Lucene index over the HomoloGene database.
It handles the mapping between a HomoloGene database entry and structured Java objects of type
<a href="api/com/aliasi/lingmed/homologene/Homologene.html"><code>HomologeneGroup</code></a>.
Each database entry contains a set of one or more genes, and these are mapped to Java objects of type
<a href="api/com/aliasi/lingmed/homologene/HomologeneGene.html"><code>HomologeneGene</code></a>.
</p>

<h3>Downloading HomoloGene</h3>

<p>
Homologene is distributed as a compressed XML file.
Like EntrezGene, Homologene is distributed as one dataset,
and each new release replaces the previous release.
It can be downloaded from NCBI vis the URL:
<a href="ftp://ftp.ncbi.nih.gov/pub/HomoloGene/current/homologene.xml.gz">
ftp://ftp.ncbi.nih.gov/pub/HomoloGene/current/homologene.xml.gz</a>.
</p>

<h3>Indexing HomoloGene</h3>

<p>
The program
<a href="api/com/aliasi/lingmed/homologene/IndexHomologene.html"><code>IndexHomologene</code></a>
processes the Homologene distribution file 
and adds each HomologeneGroup entry to a Lucene Index on the local filesystem.
It takes the following arguments:
</p>

<p>
<ul>
<li>
<b>index</b>
&mdash;  Path to the Lucene index file. 
</li>
<li>
<b>distFile</b>
&mdash; Path to the XML file.
</li>
</ul>
</p>

<p>
The LingMed 
<a href="../build.xml">build.xml</a> file
defines properties and tasks to run this program:
</p>

<pre class="code">
&lt;property name="homologene.lucene.index.dir"
          value="luceneHomologene"/&gt;
&lt;property name="homologene.dist.file"
          value="C:\\aliasi\data\homologene\homologene.xml"/&gt;
...
&lt;target name="indexHomologene"
        description="create Lucene HomologeneGene index"
        depends="jar"&gt;
  &lt;java classname="com.aliasi.lingmed.entrezgene.IndexHomologene"
        maxMemory="800G"
        fork="true"&gt;
    &lt;classpath refid="classpath.standard"/&gt;
    &lt;arg value="-index=${homologene.lucene.index.dir}"/&gt;
    &lt;arg value="-distFile=${homologene.dist.file}"/&gt;
  &lt;/java&gt;
&lt;/target&gt;
</pre>

<p>
The implementation of <code>IndexHomologene</code> parallels that
of <code>IndexEntrezGene</code>.
</p>

<h3>Searching HomoloGene</h3>

<p>
The interface
<a href="api/com/aliasi/lingmed/homologene/HomologeneSearcher.html"><code>HomologeneSearcher</code></a>
defines the search functionality for Lucene index of HomoloGene.
</p>


<a name="omim"><h2>OMIM</h2></a>

<p>
The package 
<a href="api/com/aliasi/lingmed/omim/package-summary.html">
<code>com.aliasi.lingmed.omim</code></a>
contains tools for creating and searching a Lucene index 
over the OMIM database.
It handles the mapping between an OMIM record and structured Java objects of type
<a href="api/com/aliasi/lingmed/omim/OmimRecord.html"><code>OmimRecord</code></a>.
</p>

<h3>Downloading OMIM</h3>

<p>
The OMIM database is available for FTP download from
<A href="ftp://ftp.ncbi.nih.gov/repository/OMIM/omim.txt.Z">
ftp.ncbi.nih.gov/repository/OMIM/omim.txt.Z</A>
It is a compressed ASCII text file
containing a set of records.
Each record contains a number of fields.
Records begin with the header: 
<code>*RECORD*</code> on a separate line.
Fields begin with the header: 
<code>*FIELD* &lt;FIELD KEY&gt;</code> also on a separate line.
</p>

<h3>Indexing OMIM</h3>

<p>
<p>
The program
<a href="api/com/aliasi/lingmed/omim/IndexOmim.html"><code>IndexOmim</code></a>
processes the OMIM distribution file 
and adds each OmimRecord to a Lucene Index on the local filesystem.
It takes the following arguments:
</p>

<p>
<ul>
<li>
<b>index</b>
&mdash;  Path to the Lucene index file. 
</li>
<li>
<b>distFile</b>
&mdash; Path to the XML file.
</li>
</ul>
</p>

<p>
The LingMed 
<a href="../build.xml">build.xml</a> file
defines properties and tasks to run this program:
</p>
 
<pre class="code">
&lt;property name="omim.lucene.index.dir"
          value="luceneOMIM"/&gt;
&lt;property name="omim.dist.file"
          value="C:\\aliasi\data\mim\omim.txt"/&gt;
...
&lt;target name="indexOmim"
        description="create Lucene OMIM index"
        depends="jar"&gt;
  &lt;java classname="com.aliasi.lingmed.omim.IndexOmim"
        maxMemory="800M"
        fork="true"&gt;
    &lt;classpath refid="classpath.standard"/&gt;
    &lt;arg value="-index=${omim.lucene.index.dir}"/&gt;
    &lt;arg value="-distFile=${omim.dist.file}"/&gt;
  &lt;/java&gt;
&lt;/target&gt;
</pre>


<p>
The class 
<a href="api/com/aliasi/lingmed/omim/OmimParser.html">
<code>OmimParser</code></a>
parses the distribution file into its constituent 
<a href="api/com/aliasi/lingmed/omim/OmimRecord.html">
<code>OmimRecord</code></a>s.
<code>OmimParser</code> extends 
<a href="http://alias-i.com/lingpipe/docs/api/com/aliasi/corpus/InputSourceParser.html">
<code>com.aliasi.corpus.InputSourceParser&lt;H extends Handler&gt;</code></a>.
It implements a <code>parse</code> method which reads in the text file
line by line, collecting the inputs into an <code>OmimRecord</code>,
which is handed off to the <code>ObjectHandler</code>'s <code>handle</code>
method.
</p>

<h3>Searching OMIM</h3>

<p>
The interface
<a href="api/com/aliasi/lingmed/omim/OmimSearcher.html"><code>OmimSearcher</code></a>
defines the search functionality for Lucene index of OMIM records.
</p>





</div><!-- content -->

<div id="foot">
<p>
&#169; 2003&ndash;2006 &nbsp;
<a href="mailto:lingpipe@alias-i.com">alias-i</a>
</p>
</div>

</body>

</html>


